packages/contracts-core/contracts/test/GovernanceMessage.t.sol

```diff
@@ -0,0 +1,488 @@
+    }
+
+    function test_isValidBatchWrongLengthFuzzed(bytes memory data) public {
+        vm.assume(data.length != 32);
+        data = abi.encodePacked(hex"01", data);
+        assertFalse(GovernanceMessage.isValidBatch(data.ref(0)));
+    }
+
+    function test_isBatchDetectsViewType() public {
+        bytes memory data = hex"01";
+        // Append an empty bytes array of 32 bytes
+        data = abi.encodePacked(data, new bytes(32));
+        assert(GovernanceMessage.isBatch(data.ref(1)));
+        assertFalse(GovernanceMessage.isBatch(data.ref(2)));
+    }
+
+    function test_isBatchDetectsViewTypeFuzzed(
+        uint8 viewType,
+        bytes memory data
+    ) public {
+        vm.assume(data.length == 32);
+        bytes memory prefix = abi.encodePacked(viewType);
+        // Append an empty bytes array of 32 bytes
+        data = abi.encodePacked(prefix, data);
+        if (viewType == 1) {
+            assert(GovernanceMessage.isBatch(data.ref(viewType)));
+        } else {
+            assertFalse(GovernanceMessage.isBatch(data.ref(viewType)));
+        }
+    }
+
+    function test_isBatchDifferentViewTypeToPrefixFuzzed(
+        uint8 viewType,
+        bytes memory data
+    ) public {
+        vm.assume(data.length == 32 && viewType < 3);
+        // the prefix is different to the type of the view
+        bytes memory prefix = hex"01";
+        // Append an empty bytes array of 32 bytes
+        data = abi.encodePacked(prefix, data);
+        if (viewType == 1) {
+            assert(GovernanceMessage.isBatch(data.ref(viewType)));
+        } else {
+            assertFalse(GovernanceMessage.isBatch(data.ref(viewType)));
+        }
+    }
+
+    function test_tryAsBatchForBatchReturnsBatch() public {
+        bytes memory data = hex"01";
+        // Appnend an empty bytes array of 32 bytes
+        data = abi.encodePacked(data, new bytes(32));
+        bytes29 dataView = data.ref(1);
+        // We compare both the type of the view and the contents of the memory location
+        // to where the view points
+        assertEq(uint256(dataView.tryAsBatch().typeOf()), 1);
+        assertEq(dataView.tryAsBatch().keccak(), dataView.keccak());
+        dataView = data.ref(34);
+        // even if instantiate the view with a differerent type, it can still
+        // be cast to a Batch type (1)
+        assertEq(uint256(dataView.tryAsBatch().typeOf()), 1);
+        assertEq(dataView.tryAsBatch().keccak(), dataView.keccak());
+    }
+
+    function test_tryAsBatchForBatchReturnsBatchFuzzed(uint40 viewType) public {
+        bytes memory data = hex"01";
+        // Append an empty bytes array of 32 bytes
+        data = abi.encodePacked(data, new bytes(32));
+        bytes29 dataView = data.ref(viewType);
+        // We compare both the type of the view and the contents of the memory location
+        // to where the view points
+        assertEq(uint256(dataView.tryAsBatch().typeOf()), 1);
```

Comment:
a lot of the type comparisons are done by casting to uint. it would be better to use `BridgeMessage.Types.TypeName` instead?